### ISSUE TITLE
BUG: SHAP does not take into account certain key parameters from the xgboost sklearn models that can affect predictions

### ISSUE THREAD

**Issue by vancromy** [shap/shap#3313]
### Issue Description
The TreeEnsemble class responsible for automatically identifying and loading in the passed model does not take into account important parameters set at the class level on a sklearn compatible xgboost model. This is because these parameters are passed to the DMatrix object by the xgboost library in the lower level functions of their sklearn API before being sent to the fit/predict methods (see here for example: https://github.com/dmlc/xgboost/blob/4d7a187cb0ebc2194abc1efe0aaee6f06af937c5/python-package/xgboost/sklearn.py#L514)

I stumbled upon this because my shap values were not exactly equal to my predicted probabilities (a check I was performing) where I calculated the probabilities from the `predict_proba` method and I calculated the shap values using the TreeExplainer and then converted the shap values to shap probabilities (since the values are output as margin and not probabilities). The reason for the difference was that I use a custom value for representing missing data in my models (i.e. I set `missing = 1e-8` on my XGBClassifier instance) and when `predict_proba` gets called this missing value argument is passed lower down to the booster object via the DMatrix. SHAP however does not propagate these class level params further to the DMatrix as can be seen on these lines: https://github.com/shap/shap/blob/5121282dfd6fb66a21a4866ed3fa6ba6917a7efa/shap/explainers/_tree.py#L357-L358

For now the solution is to build the DMatrix up front before passing it to the explainer but I wanted to raise this as an issue because its a very silent bug and if users are not careful they will actually get the wrong shap values for their model.

These lines of the TreeEnsemble class would potentially need to be changed to account for reading in these parameters:
https://github.com/shap/shap/blob/5121282dfd6fb66a21a4866ed3fa6ba6917a7efa/shap/explainers/_tree.py#L971-L1015

### Minimal Reproducible Example
```python

import pandas as pd
import xgboost
import shap

df = pd.DataFrame(your_data)

X, y, X_val, y_val = train_test_split(df[features], df[label], 0.3)

# Assuming this is a binary classification problem)
clf = xgboost.XGBClassifier(missing=1e-8)
clf.fit(X,y)
pred_probs = clf.predict_proba(X_val)[:,1]

explainer = shap.TreeExplainer(clf)
shap_values = explainer.shap_values(X_val)

pred_probs != sigmoid(explainer.expected_value + shap_values.sum(axis=1))
```
In above snippet if you build the DMatrix upfront and pass it to the shap_values call, the shap probabilities are equal to the predicted probabilities

### Expected Behavior
I would expect the sum of the calculated shap values to be equal to the predictions of the model (in both margin and probability terms). The `assert_additivity` method of shap only asserts to a 1e-2 order of magnitude that the margin predictions are correct but if additivity had been asserted on the probaibilities this would have been picked up earlier on.

### Bug report checklist

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [x] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [x] I'd be interested in making a PR to fix this bug

### Installed versions

Clone of the master branch of the repo

**Comment by CloseChoice** [comment]
Thanks for the bug report. I have tweaked your example so that it is reproducible (the dataset and definition of sigmoid was missing):

```python
import xgboost
import shap
import numpy as np
from sklearn.model_selection import train_test_split

X, y = shap.datasets.adult()

def sigmoid(x):
  return 1 / (1 + np.exp(-x))

def logit(x):
  return np.log(x / (1 - x))

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.3)

# Assuming this is a binary classification problem)
clf = xgboost.XGBClassifier(missing=1e-8)
clf.fit(X_train, y_train)
pred_probs = clf.predict_proba(X_val)[:,1]

explainer = shap.TreeExplainer(clf)
shap_values = explainer.shap_values(X_val)

np.max(pred_probs - sigmoid(explainer.expected_value + shap_values.sum(axis=1)))
# results in 2.9802322e-07
np.max(logit(pred_probs) - explainer.expected_value - shap_values.sum(axis=1))
# results in 0.0005540848
```
Hmm, so we are within the margin for the probability. Could you provide an example where this margin is broken?

**Comment by vancromy** [comment]
> Thanks for the bug report. I have tweaked your example so that it is reproducible (the dataset and definition of sigmoid was missing):
> 
> <details>
> ```python
> import xgboost
> import shap
> import numpy as np
> from sklearn.model_selection import train_test_split
> 
> X, y = shap.datasets.adult()
> 
> def sigmoid(x):
>   return 1 / (1 + np.exp(-x))
> 
> def logit(x):
>   return np.log(x / (1 - x))
> 
> X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.3)
> 
> # Assuming this is a binary classification problem)
> clf = xgboost.XGBClassifier(missing=1e-8)
> clf.fit(X_train, y_train)
> pred_probs = clf.predict_proba(X_val)[:,1]
> 
> explainer = shap.TreeExplainer(clf)
> shap_values = explainer.shap_values(X_val)
> 
> np.max(pred_probs - sigmoid(explainer.expected_value + shap_values.sum(axis=1)))
> # results in 2.9802322e-07
> np.max(logit(pred_probs) - explainer.expected_value - shap_values.sum(axis=1))
> # results in 0.0005540848
> ```
> </details>
> Hmm, so we are within the margin for the probability. Could you provide an example where this margin is broken?

@CloseChoice Thanks for looking at this and providing an example dataset. Your example did not have any missing data so I added some to better reflect the bug :)

```python
import xgboost
import shap
import numpy as np
from sklearn.model_selection import train_test_split

X, y = shap.datasets.adult()


def sigmoid(x):
    return 1 / (1 + np.exp(-x))


def logit(x):
    return np.log(x / (1 - x))


# Randomly set some features of X to NaN
X_nan = X.copy()
X_nan.loc[
    X_nan.sample(frac=0.3, random_state=42).index,
    X_nan.columns.to_series().sample(frac=0.5, random_state=42),
] = 1e-8


X_train, X_val, y_train, y_val = train_test_split(X_nan, y, test_size=0.3)

# Assuming this is a binary classification problem)
clf = xgboost.XGBClassifier(missing=1e-8, random_state=42)
clf.fit(X_train, y_train)
pred_probs = clf.predict_proba(X_val)[:, 1]

# With bug
explainer = shap.TreeExplainer(clf)
shap_values_bug = explainer.shap_values(X_val)
print(
    np.max(pred_probs - sigmoid(explainer.expected_value + shap_values_bug.sum(axis=1)))
)
# results in 0.88415104

# Without bug (suggested fix of persisting the params to the DMatrix)
X_val_d = xgboost.DMatrix(X_val, missing=clf.missing, nthread=clf.n_jobs)
explainer = shap.TreeExplainer(clf)
shap_values = explainer.shap_values(X_val_d)
print(np.max(pred_probs - sigmoid(explainer.expected_value + shap_values.sum(axis=1))))
# results in 2.682209e-07
```

**Comment by CloseChoice** [comment]
Thanks for the example. To me your suggested fix sounds fine. I am not a maintainer but I see that you are interested in writing a PR to fix this, could you just implement the solution you proposed (maybe a bit more generic, so we find all DMatrix kwargs and hand it to the DMatrix) and add this example as a test case?

We could also ping the maintainers to get their view on this.

**Comment by vancromy** [comment]
> Thanks for the example. To me your suggested fix sounds fine. I see that you are interested in writing a PR to fix this, could you just implement the solution you proposed (maybe a bit more generic, so we find all DMatrix kwargs and hand it to the DMatrix) and add this example as a test case?

Sure thing. I already started the work. Should have a PR up in a couple hours :)
